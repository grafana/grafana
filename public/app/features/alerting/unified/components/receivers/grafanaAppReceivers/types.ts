import { SupportedPlugin } from '../../../types/pluginBridges';

export const GRAFANA_APP_RECEIVERS_SOURCE_IMAGE: Record<SupportedPlugin, string> = {
  [SupportedPlugin.OnCall]: 'public/img/alerting/oncall_logo.svg',

  [SupportedPlugin.Incident]: '',
  [SupportedPlugin.MachineLearning]: '',
  [SupportedPlugin.Labels]: '',
  [SupportedPlugin.Slo]: '',
};
