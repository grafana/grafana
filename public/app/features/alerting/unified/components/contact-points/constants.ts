export const RECEIVER_STATUS_KEY = Symbol('receiver_status');
export const RECEIVER_META_KEY = Symbol('receiver_metadata');
export const RECEIVER_PLUGIN_META_KEY = Symbol('receiver_plugin_metadata');
