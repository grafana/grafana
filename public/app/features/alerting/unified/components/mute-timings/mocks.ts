import { AlertManagerCortexConfig, MuteTimeInterval } from 'app/plugins/datasource/alertmanager/types';

export const muteTimeInterval: MuteTimeInterval = {
  name: 'default-mute',
  time_intervals: [
    {
      times: [
        {
          start_time: '12:00',
          end_time: '24:00',
        },
      ],
      days_of_month: ['15', '-1'],
      months: ['august:december', 'march'],
    },
  ],
};
/** Alertmanager config where time intervals are stored in `mute_time_intervals` property */
export const defaultConfig: AlertManagerCortexConfig = {
  alertmanager_config: {
    receivers: [{ name: 'default' }, { name: 'critical' }],
    route: {
      receiver: 'default',
      group_by: ['alertname'],
      routes: [
        {
          matchers: ['env=prod', 'region!=EU'],
          mute_time_intervals: [muteTimeInterval.name],
        },
      ],
    },
    templates: [],
    mute_time_intervals: [muteTimeInterval],
  },
  template_files: {},
};
