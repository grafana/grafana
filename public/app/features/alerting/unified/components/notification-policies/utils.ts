import { ControllerRenderProps } from 'react-hook-form';

import { SelectableValue } from '@grafana/data';
import { ContactPointWithMetadata } from 'app/features/alerting/unified/components/contact-points/utils';

export const handleContactPointSelect = (
  value: SelectableValue<ContactPointWithMetadata>,
  onChange: ControllerRenderProps['onChange']
) => {
  if (value === null) {
    return onChange(null);
  }

  if (!value) {
    return onChange('');
  }

  return onChange(value.value?.name);
};
