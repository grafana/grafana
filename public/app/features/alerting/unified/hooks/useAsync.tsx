/**
 * Copied from https://github.com/react-hookz/web/blob/579a445fcc9f4f4bb5b9d5e670b2e57448b4ee50/src/useAsync/index.ts
 */
import { useMemo, useRef, useState } from 'react';

import { stringifyErrorLike } from '../utils/misc';

export type AsyncStatus = 'loading' | 'success' | 'error' | 'not-executed';

export type AsyncState<Result> =
  | {
      status: 'not-executed';
      error: undefined;
      result: Result;
    }
  | {
      status: 'success';
      error: undefined;
      result: Result;
    }
  | {
      status: 'error';
      error: Error;
      result: Result;
    }
  | {
      status: AsyncStatus;
      error: Error | undefined;
      result: Result;
    };

export type UseAsyncActions<Result, Args extends unknown[] = unknown[]> = {
  /**
   * Reset state to initial.
   */
  reset: () => void;
  /**
   * Execute the async function manually.
   */
  execute: (...args: Args) => Promise<Result>;
};

export type UseAsyncMeta<Result, Args extends unknown[] = unknown[]> = {
  /**
   * Latest promise returned from the async function.
   */
  promise: Promise<Result> | undefined;
  /**
   * List of arguments applied to the latest async function invocation.
   */
  lastArgs: Args | undefined;
};

export function useAsync<Result, Args extends unknown[] = unknown[]>(
  asyncFn: (...params: Args) => Promise<Result>,
  initialValue: Result
): [UseAsyncActions<Result, Args>, AsyncState<Result>, UseAsyncMeta<Result, Args>];
export function useAsync<Result, Args extends unknown[] = unknown[]>(
  asyncFn: (...params: Args) => Promise<Result>,
  initialValue?: Result
): [UseAsyncActions<Result, Args>, AsyncState<Result | undefined>, UseAsyncMeta<Result, Args>];

/**
 * Tracks the result and errors of the provided async function and provides handles to control its execution.
 *
 * @param asyncFn Function that returns a promise.
 * @param initialValue Value that will be set on initialisation before the async function is
 * executed.
 */
export function useAsync<Result, Args extends unknown[] = unknown[]>(
  asyncFn: (...params: Args) => Promise<Result>,
  initialValue?: Result
): [UseAsyncActions<Result, Args>, AsyncState<Result | undefined>, UseAsyncMeta<Result, Args>] {
  const [state, setState] = useState<AsyncState<Result | undefined>>({
    status: 'not-executed',
    error: undefined,
    result: initialValue,
  });
  const promiseRef = useRef<Promise<Result>>();
  const argsRef = useRef<Args>();

  const methods = useSyncedRef({
    execute(...params: Args) {
      argsRef.current = params;
      const promise = asyncFn(...params);
      promiseRef.current = promise;

      setState((s) => ({ ...s, status: 'loading' }));

      promise.then(
        (result) => {
          // We dont want to handle result/error of non-latest function
          // this approach helps to avoid race conditions

          if (promise === promiseRef.current) {
            setState((s) => ({ ...s, status: 'success', error: undefined, result }));
          }
        },
        (error: Error) => {
          // We dont want to handle result/error of non-latest function
          // this approach helps to avoid race conditions
          if (promise === promiseRef.current) {
            setState((s) => ({ ...s, status: 'error', error }));
          }
        }
      );

      return promise;
    },
    reset() {
      setState({
        status: 'not-executed',
        error: undefined,
        result: initialValue,
      });
      promiseRef.current = undefined;
      argsRef.current = undefined;
    },
  });

  return [
    useMemo(
      () => ({
        reset() {
          methods.current.reset();
        },
        execute: (...params: Args) => methods.current.execute(...params),
      }),
      // eslint-disable-next-line react-hooks/exhaustive-deps
      []
    ),
    state,
    { promise: promiseRef.current, lastArgs: argsRef.current },
  ];
}

/**
 * Like `useRef`, but it returns immutable ref that contains actual value.
 *
 * @param value
 */
function useSyncedRef<T>(value: T): { readonly current: T } {
  const ref = useRef(value);

  ref.current = value;

  return useMemo(
    () =>
      Object.freeze({
        get current() {
          return ref.current;
        },
      }),
    []
  );
}

// --- utility functions to help with request state assertions ---

export function isError<T>(state: AsyncState<T>) {
  return state.status === 'error';
}

export function isSuccess<T>(state: AsyncState<T>) {
  return state.status === 'success';
}

export function isUninitialized<T>(state: AsyncState<T>) {
  return state.status === 'not-executed';
}

export function isLoading<T>(state: AsyncState<T>) {
  return state.status === 'loading';
}

export function anyOfRequestState(...states: Array<AsyncState<unknown>>) {
  return {
    uninitialized: states.every(isUninitialized),
    loading: states.some(isLoading),
    error: states.find(isError)?.error,
    success: states.some(isSuccess),
  };
}

/**
 * This is only used for testing and serializing the async state
 */
export function SerializeState({ state }: { state: AsyncState<unknown> }) {
  return (
    <>
      {isUninitialized(state) && 'uninitialized'}
      {isLoading(state) && 'loading'}
      {isSuccess(state) && 'success'}
      {isSuccess(state) && `result: ${JSON.stringify(state.result, null, 2)}`}
      {isError(state) && `error: ${stringifyErrorLike(state.error)}`}
    </>
  );
}
