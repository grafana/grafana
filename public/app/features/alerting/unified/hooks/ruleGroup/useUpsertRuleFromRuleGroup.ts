import { produce } from 'immer';
import { isEqual } from 'lodash';

import { t } from 'app/core/internationalization';
import { dispatch } from 'app/store/store';
import { RuleGroupIdentifier, EditableRuleIdentifier } from 'app/types/unified-alerting';
import { PostableRuleDTO } from 'app/types/unified-alerting-dto';

import { alertRuleApi } from '../../api/alertRuleApi';
import { addRuleAction, updateRuleAction } from '../../reducers/ruler/ruleGroups';
import { fetchRulerRulesAction } from '../../state/actions';
import { isGrafanaRuleIdentifier, isGrafanaRulerRule } from '../../utils/rules';
import { useAsync } from '../useAsync';

import { useDeleteRuleFromGroup } from './useDeleteRuleFromGroup';
import { useProduceNewRuleGroup } from './useProduceNewRuleGroup';

/**
 * This hook will add a single rule to a rule group â€“ a new rule group will be created if it does not already exist.
 */
export function useAddRuleToRuleGroup() {
  const [produceNewRuleGroup] = useProduceNewRuleGroup();
  const [upsertRuleGroup] = alertRuleApi.endpoints.upsertRuleGroupForNamespace.useMutation();

  const successMessage = t('alerting.rules.add-rule.success', 'Rule added successfully');

  return useAsync(async (ruleGroup: RuleGroupIdentifier, rule: PostableRuleDTO, interval?: string) => {
    const { namespaceName, dataSourceName } = ruleGroup;

    // the new rule might have to be created in a new group, pass name and interval (optional) to the action
    const action = addRuleAction({ rule, interval, groupName: ruleGroup.groupName });
    const { newRuleGroupDefinition, rulerConfig } = await produceNewRuleGroup(ruleGroup, action);

    const result = upsertRuleGroup({
      rulerConfig,
      namespace: namespaceName,
      payload: newRuleGroupDefinition,
      notificationOptions: { successMessage },
    }).unwrap();

    // @TODO remove
    await dispatch(fetchRulerRulesAction({ rulesSourceName: dataSourceName }));

    return result;
  });
}

/**
 * This hook will update an existing rule within a rule group, does not support moving the rule to another namespace / group
 */
export function useUpdateRuleInRuleGroup() {
  const [produceNewRuleGroup] = useProduceNewRuleGroup();
  const [moveRuleToGroup] = useMoveRuleToRuleGroup();
  const [upsertRuleGroup] = alertRuleApi.endpoints.upsertRuleGroupForNamespace.useMutation();

  const successMessage = t('alerting.rules.update-rule.success', 'Rule updated successfully');

  return useAsync(
    async (
      ruleGroup: RuleGroupIdentifier,
      ruleIdentifier: EditableRuleIdentifier,
      ruleDefinition: PostableRuleDTO,
      targetRuleGroup?: RuleGroupIdentifier
    ) => {
      const { namespaceName } = ruleGroup;
      const finalRuleDefinition = copyGrafanaUID(ruleIdentifier, ruleDefinition);

      // check if the existing rule and the form values have the same rule group identifier
      const sameTargetRuleGroup = isEqual(ruleGroup, targetRuleGroup);
      if (targetRuleGroup && !sameTargetRuleGroup) {
        const result = moveRuleToGroup.execute(ruleGroup, targetRuleGroup, ruleIdentifier, ruleDefinition);
        return result;
      }

      const action = updateRuleAction({ identifier: ruleIdentifier, rule: finalRuleDefinition });
      const { newRuleGroupDefinition, rulerConfig } = await produceNewRuleGroup(ruleGroup, action);

      return upsertRuleGroup({
        rulerConfig,
        namespace: namespaceName,
        payload: newRuleGroupDefinition,
        notificationOptions: { successMessage },
      }).unwrap();
    }
  );
}

/**
 * This hook will move an existing rule to another namespace or group. The rule definition can also be modified.
 * For Grafana-managed rules we can perform a single atomic move operation by copying the rule UID from the previous rule definition.
 */
export function useMoveRuleToRuleGroup() {
  const [produceNewRuleGroup] = useProduceNewRuleGroup();
  const [deleteRuleFromGroup] = useDeleteRuleFromGroup();
  const [upsertRuleGroup] = alertRuleApi.endpoints.upsertRuleGroupForNamespace.useMutation();

  const successMessage = t('alerting.rules.update-rule.success', 'Rule updated successfully');

  return useAsync(
    async (
      currentRuleGroup: RuleGroupIdentifier,
      targetRuleGroup: RuleGroupIdentifier,
      ruleIdentifier: EditableRuleIdentifier,
      ruleDefinition: PostableRuleDTO
    ) => {
      const finalRuleDefinition = copyGrafanaUID(ruleIdentifier, ruleDefinition);

      // 1. add the rule to the new namespace / group / ruler target
      const addRuleToGroup = addRuleAction({ rule: finalRuleDefinition });
      const { newRuleGroupDefinition: newTargetGroup, rulerConfig: targetGroupRulerConfig } = await produceNewRuleGroup(
        targetRuleGroup,
        addRuleToGroup
      );

      const result = await upsertRuleGroup({
        rulerConfig: targetGroupRulerConfig,
        namespace: targetRuleGroup.namespaceName,
        payload: newTargetGroup,
        notificationOptions: { successMessage },
      }).unwrap();

      // 2. if not Grafana-managed: remove the rule from the existing namespace / group / ruler
      if (!isGrafanaRuleIdentifier(ruleIdentifier)) {
        await deleteRuleFromGroup.execute(currentRuleGroup, ruleIdentifier);
      }

      return result;
    }
  );
}

function copyGrafanaUID(ruleIdentifier: EditableRuleIdentifier, ruleDefinition: PostableRuleDTO) {
  const isGrafanaManagedRuleIdentifier = isGrafanaRuleIdentifier(ruleIdentifier);

  // by copying over the rule UID the backend will perform an atomic move operation
  // so there is no need for us to manually remove it from the previous group
  return produce(ruleDefinition, (draft) => {
    const isGrafanaManagedRuleDefinition = isGrafanaRulerRule(draft);

    if (isGrafanaManagedRuleIdentifier && isGrafanaManagedRuleDefinition) {
      draft.grafana_alert.uid = ruleIdentifier.uid;
    }
  });
}
