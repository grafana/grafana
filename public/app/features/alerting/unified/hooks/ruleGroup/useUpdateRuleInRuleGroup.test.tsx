import { produce } from 'immer';
import { render } from 'test/test-utils';
import { byRole, byText } from 'testing-library-selector';

import { AccessControlAction } from 'app/types/accessControl';
import { EditableRuleIdentifier, GrafanaRuleIdentifier, RuleGroupIdentifier } from 'app/types/unified-alerting';
import { PostableRuleDTO } from 'app/types/unified-alerting-dto';

import { setupMswServer } from '../../mockApi';
import { grantUserPermissions } from '../../mocks';
import {
  grafanaRulerGroupName,
  grafanaRulerGroupName2,
  grafanaRulerNamespace,
  grafanaRulerRule,
} from '../../mocks/grafanaRulerApi';
import { NAMESPACE_1, group1 } from '../../mocks/mimirRulerApi';
import { mimirDataSource } from '../../mocks/server/configure';
import { MIMIR_DATASOURCE_UID } from '../../mocks/server/constants';
import { captureRequests, serializeRequests } from '../../mocks/server/events';
import { GRAFANA_RULES_SOURCE_NAME } from '../../utils/datasource';
import { fromRulerRuleAndRuleGroupIdentifier } from '../../utils/rule-id';
import { SerializeState } from '../useAsync';

import { useUpdateRuleInRuleGroup } from './useUpsertRuleFromRuleGroup';

setupMswServer();

beforeAll(() => {
  grantUserPermissions([
    AccessControlAction.AlertingRuleExternalRead,
    AccessControlAction.AlertingRuleExternalWrite,
    AccessControlAction.AlertingRuleRead,
    AccessControlAction.AlertingRuleCreate,
  ]);
});

describe('Updating a Grafana managed rule', () => {
  it('should update a rule in an existing group', async () => {
    const capture = captureRequests((r) => r.method === 'POST');

    const ruleGroupID: RuleGroupIdentifier = {
      dataSourceName: GRAFANA_RULES_SOURCE_NAME,
      groupName: grafanaRulerGroupName,
      namespaceName: grafanaRulerNamespace.uid,
    };

    const ruleID: GrafanaRuleIdentifier = {
      ruleSourceName: GRAFANA_RULES_SOURCE_NAME,
      uid: grafanaRulerRule.grafana_alert.uid,
    };

    const newRule = produce(grafanaRulerRule, (draft) => {
      draft.grafana_alert.title = 'updated rule title';
    });

    const { user } = render(
      <UpdateRuleTestComponent ruleGroupIdentifier={ruleGroupID} ruleID={ruleID} rule={newRule} />
    );
    await user.click(byRole('button').get());

    expect(await byText(/success/i).find()).toBeInTheDocument();

    const requests = await capture;
    const serializedRequests = await serializeRequests(requests);
    expect(serializedRequests).toMatchSnapshot();
  });

  it('should move a rule in to another group', async () => {
    const capture = captureRequests((r) => r.method === 'POST');

    const ruleGroupID: RuleGroupIdentifier = {
      dataSourceName: GRAFANA_RULES_SOURCE_NAME,
      groupName: grafanaRulerGroupName,
      namespaceName: grafanaRulerNamespace.uid,
    };

    const targetRuleGroupID: RuleGroupIdentifier = {
      dataSourceName: GRAFANA_RULES_SOURCE_NAME,
      groupName: grafanaRulerGroupName2,
      namespaceName: grafanaRulerNamespace.uid,
    };

    const ruleID: GrafanaRuleIdentifier = {
      ruleSourceName: GRAFANA_RULES_SOURCE_NAME,
      uid: grafanaRulerRule.grafana_alert.uid,
    };

    const newRule = produce(grafanaRulerRule, (draft) => {
      draft.grafana_alert.title = 'updated rule title';
    });

    const { user } = render(
      <UpdateRuleTestComponent
        ruleGroupIdentifier={ruleGroupID}
        targetRuleGroupIdentifier={targetRuleGroupID}
        ruleID={ruleID}
        rule={newRule}
      />
    );
    await user.click(byRole('button').get());

    expect(await byText(/success/i).find()).toBeInTheDocument();

    const requests = await capture;
    const serializedRequests = await serializeRequests(requests);
    expect(serializedRequests).toMatchSnapshot();
  });

  it('should fail if the rule does not exist in the group', async () => {
    const ruleGroupID: RuleGroupIdentifier = {
      dataSourceName: GRAFANA_RULES_SOURCE_NAME,
      groupName: grafanaRulerGroupName,
      namespaceName: grafanaRulerNamespace.uid,
    };

    const ruleID: GrafanaRuleIdentifier = {
      ruleSourceName: GRAFANA_RULES_SOURCE_NAME,
      uid: 'does-not-exist',
    };

    const newRule = produce(grafanaRulerRule, (draft) => {
      draft.grafana_alert.title = 'updated rule title';
    });

    const { user } = render(
      <UpdateRuleTestComponent ruleGroupIdentifier={ruleGroupID} ruleID={ruleID} rule={newRule} />
    );
    await user.click(byRole('button').get());

    expect(await byText(/error: no rule matching identifier found/i).find()).toBeInTheDocument();
  });

  it('should fail if the rule group does not exist', async () => {
    const ruleGroupID: RuleGroupIdentifier = {
      dataSourceName: GRAFANA_RULES_SOURCE_NAME,
      groupName: 'does-not-exist',
      namespaceName: 'does-not-exist',
    };

    const ruleID: GrafanaRuleIdentifier = {
      ruleSourceName: GRAFANA_RULES_SOURCE_NAME,
      uid: 'does-not-exist',
    };

    const newRule = produce(grafanaRulerRule, (draft) => {
      draft.grafana_alert.title = 'updated rule title';
    });

    const { user } = render(
      <UpdateRuleTestComponent ruleGroupIdentifier={ruleGroupID} ruleID={ruleID} rule={newRule} />
    );
    await user.click(byRole('button').get());

    expect(await byText(/error/i).find()).toBeInTheDocument();
  });
});

describe('Updating a Data source managed rule', () => {
  beforeEach(() => {
    mimirDataSource();
  });

  it('should update a rule in an existing group', async () => {
    const capture = captureRequests((r) => r.method === 'POST');

    const groupToUpdate = group1;
    const ruleToUpdate = groupToUpdate.rules[0];

    const ruleGroupID: RuleGroupIdentifier = {
      dataSourceName: MIMIR_DATASOURCE_UID,
      groupName: groupToUpdate.name,
      namespaceName: NAMESPACE_1,
    };

    const ruleID = fromRulerRuleAndRuleGroupIdentifier(ruleGroupID, ruleToUpdate);

    const newRule = produce(grafanaRulerRule, (draft) => {
      draft.grafana_alert.title = 'updated rule title';
    });

    const { user } = render(
      <UpdateRuleTestComponent ruleGroupIdentifier={ruleGroupID} ruleID={ruleID} rule={newRule} />
    );
    await user.click(byRole('button').get());

    expect(await byText(/success/i).find()).toBeInTheDocument();

    const requests = await capture;
    const serializedRequests = await serializeRequests(requests);
    expect(serializedRequests).toMatchSnapshot();
  });

  it('should be able to move a rule if target group is different from current group', async () => {
    const capture = captureRequests((r) => r.method === 'POST' || r.method === 'DELETE');

    const groupToUpdate = group1;
    const ruleToUpdate = groupToUpdate.rules[0];

    const ruleGroupID: RuleGroupIdentifier = {
      dataSourceName: MIMIR_DATASOURCE_UID,
      groupName: groupToUpdate.name,
      namespaceName: NAMESPACE_1,
    };

    const targetRuleGroupID: RuleGroupIdentifier = {
      dataSourceName: MIMIR_DATASOURCE_UID,
      groupName: 'a new group',
      namespaceName: NAMESPACE_1,
    };

    const ruleID = fromRulerRuleAndRuleGroupIdentifier(ruleGroupID, ruleToUpdate);

    const newRule = produce(grafanaRulerRule, (draft) => {
      draft.grafana_alert.title = 'updated rule title';
    });

    const { user } = render(
      <UpdateRuleTestComponent
        ruleGroupIdentifier={ruleGroupID}
        targetRuleGroupIdentifier={targetRuleGroupID}
        ruleID={ruleID}
        rule={newRule}
      />
    );
    await user.click(byRole('button').get());

    expect(await byText(/success/i).find()).toBeInTheDocument();

    const requests = await capture;
    const serializedRequests = await serializeRequests(requests);
    expect(serializedRequests).toMatchSnapshot();
  });

  it('should fail if the rule does not exist in the group', async () => {
    const groupToUpdate = group1;

    const ruleGroupID: RuleGroupIdentifier = {
      dataSourceName: MIMIR_DATASOURCE_UID,
      groupName: groupToUpdate.name,
      namespaceName: NAMESPACE_1,
    };

    const newRule = produce(grafanaRulerRule, (draft) => {
      draft.grafana_alert.title = 'updated rule title';
    });

    const ruleID = fromRulerRuleAndRuleGroupIdentifier(ruleGroupID, newRule);

    const { user } = render(
      <UpdateRuleTestComponent ruleGroupIdentifier={ruleGroupID} ruleID={ruleID} rule={newRule} />
    );
    await user.click(byRole('button').get());

    expect(await byText(/error: no rule matching identifier found/i).find()).toBeInTheDocument();
  });

  it('should fail if the rule group does not exist', async () => {
    const groupToUpdate = group1;
    const ruleToUpdate = groupToUpdate.rules[0];

    const ruleGroupID: RuleGroupIdentifier = {
      dataSourceName: MIMIR_DATASOURCE_UID,
      groupName: 'does-not-exist',
      namespaceName: NAMESPACE_1,
    };

    const ruleID = fromRulerRuleAndRuleGroupIdentifier(ruleGroupID, ruleToUpdate);

    const newRule = produce(grafanaRulerRule, (draft) => {
      draft.grafana_alert.title = 'updated rule title';
    });

    const { user } = render(
      <UpdateRuleTestComponent ruleGroupIdentifier={ruleGroupID} ruleID={ruleID} rule={newRule} />
    );
    await user.click(byRole('button').get());

    expect(await byText(/error/i).find()).toBeInTheDocument();
  });
});

type UpdateRuleTestComponentProps = {
  ruleGroupIdentifier: RuleGroupIdentifier;
  targetRuleGroupIdentifier?: RuleGroupIdentifier;
  ruleID: EditableRuleIdentifier;
  rule: PostableRuleDTO;
};

const UpdateRuleTestComponent = ({
  ruleGroupIdentifier,
  targetRuleGroupIdentifier,
  ruleID,
  rule,
}: UpdateRuleTestComponentProps) => {
  const [updateRule, requestState] = useUpdateRuleInRuleGroup();

  const onClick = () => {
    updateRule.execute(ruleGroupIdentifier, ruleID, rule, targetRuleGroupIdentifier);
  };

  return (
    <>
      <button onClick={() => onClick()} />
      <SerializeState state={requestState} />
    </>
  );
};
