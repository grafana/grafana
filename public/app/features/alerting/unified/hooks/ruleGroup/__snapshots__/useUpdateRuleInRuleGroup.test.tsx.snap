// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Updating a Data source managed rule should be able to move a rule if target group is different from current group 1`] = `
[
  {
    "body": {
      "name": "a new group",
      "rules": [
        {
          "annotations": {
            "summary": "Test alert",
          },
          "for": "5m",
          "grafana_alert": {
            "condition": "A",
            "data": [
              {
                "datasourceUid": "datasource-uid",
                "model": {
                  "datasource": {
                    "type": "prometheus",
                    "uid": "datasource-uid",
                  },
                  "expression": "vector(1)",
                  "queryType": "alerting",
                  "refId": "A",
                },
                "queryType": "alerting",
                "refId": "A",
                "relativeTimeRange": {
                  "from": 1000,
                  "to": 2000,
                },
              },
            ],
            "exec_err_state": "Error",
            "is_paused": false,
            "namespace_uid": "uuid020c61ef",
            "no_data_state": "NoData",
            "rule_group": "grafana-group-1",
            "title": "updated rule title",
            "uid": "4d7125fee983",
          },
          "labels": {
            "region": "nasa",
            "severity": "critical",
          },
        },
      ],
    },
    "headers": [
      [
        "content-type",
        "application/json",
      ],
      [
        "accept",
        "application/json, text/plain, */*",
      ],
    ],
    "method": "POST",
    "url": "http://localhost/api/ruler/mimir/api/v1/rules/namespace-1?subtype=mimir",
  },
  {
    "body": "",
    "headers": [
      [
        "accept",
        "application/json, text/plain, */*",
      ],
    ],
    "method": "DELETE",
    "url": "http://localhost/api/ruler/mimir/api/v1/rules/namespace-1/group-1?subtype=mimir",
  },
]
`;

exports[`Updating a Data source managed rule should update a rule in an existing group 1`] = `
[
  {
    "body": {
      "interval": "1m",
      "name": "group-1",
      "rules": [
        {
          "annotations": {
            "summary": "Test alert",
          },
          "for": "5m",
          "grafana_alert": {
            "condition": "A",
            "data": [
              {
                "datasourceUid": "datasource-uid",
                "model": {
                  "datasource": {
                    "type": "prometheus",
                    "uid": "datasource-uid",
                  },
                  "expression": "vector(1)",
                  "queryType": "alerting",
                  "refId": "A",
                },
                "queryType": "alerting",
                "refId": "A",
                "relativeTimeRange": {
                  "from": 1000,
                  "to": 2000,
                },
              },
            ],
            "exec_err_state": "Error",
            "is_paused": false,
            "namespace_uid": "uuid020c61ef",
            "no_data_state": "NoData",
            "rule_group": "grafana-group-1",
            "title": "updated rule title",
            "uid": "4d7125fee983",
          },
          "labels": {
            "region": "nasa",
            "severity": "critical",
          },
        },
      ],
    },
    "headers": [
      [
        "content-type",
        "application/json",
      ],
      [
        "accept",
        "application/json, text/plain, */*",
      ],
    ],
    "method": "POST",
    "url": "http://localhost/api/ruler/mimir/api/v1/rules/namespace-1?subtype=mimir",
  },
]
`;

exports[`Updating a Grafana managed rule should move a rule in to another group 1`] = `
[
  {
    "body": {
      "interval": "1m",
      "name": "grafana-group-2",
      "rules": [
        {
          "annotations": {
            "summary": "Test alert",
          },
          "for": "5m",
          "grafana_alert": {
            "condition": "A",
            "data": [
              {
                "datasourceUid": "datasource-uid",
                "model": {
                  "datasource": {
                    "type": "prometheus",
                    "uid": "datasource-uid",
                  },
                  "expression": "vector(1)",
                  "queryType": "alerting",
                  "refId": "A",
                },
                "queryType": "alerting",
                "refId": "A",
                "relativeTimeRange": {
                  "from": 1000,
                  "to": 2000,
                },
              },
            ],
            "exec_err_state": "Error",
            "is_paused": false,
            "namespace_uid": "uuid020c61ef",
            "no_data_state": "NoData",
            "rule_group": "grafana-group-1",
            "title": "Grafana-rule",
            "uid": "4d7125fee983",
          },
          "labels": {
            "region": "nasa",
            "severity": "critical",
          },
        },
        {
          "annotations": {
            "summary": "Test alert",
          },
          "for": "5m",
          "grafana_alert": {
            "condition": "A",
            "data": [
              {
                "datasourceUid": "datasource-uid",
                "model": {
                  "datasource": {
                    "type": "prometheus",
                    "uid": "datasource-uid",
                  },
                  "expression": "vector(1)",
                  "queryType": "alerting",
                  "refId": "A",
                },
                "queryType": "alerting",
                "refId": "A",
                "relativeTimeRange": {
                  "from": 1000,
                  "to": 2000,
                },
              },
            ],
            "exec_err_state": "Error",
            "is_paused": false,
            "namespace_uid": "uuid020c61ef",
            "no_data_state": "NoData",
            "rule_group": "grafana-group-1",
            "title": "updated rule title",
            "uid": "4d7125fee983",
          },
          "labels": {
            "region": "nasa",
            "severity": "critical",
          },
        },
      ],
    },
    "headers": [
      [
        "content-type",
        "application/json",
      ],
      [
        "accept",
        "application/json, text/plain, */*",
      ],
    ],
    "method": "POST",
    "url": "http://localhost/api/ruler/grafana/api/v1/rules/uuid020c61ef?subtype=cortex",
  },
]
`;

exports[`Updating a Grafana managed rule should update a rule in an existing group 1`] = `
[
  {
    "body": {
      "interval": "1m",
      "name": "grafana-group-1",
      "rules": [
        {
          "annotations": {
            "summary": "Test alert",
          },
          "for": "5m",
          "grafana_alert": {
            "condition": "A",
            "data": [
              {
                "datasourceUid": "datasource-uid",
                "model": {
                  "datasource": {
                    "type": "prometheus",
                    "uid": "datasource-uid",
                  },
                  "expression": "vector(1)",
                  "queryType": "alerting",
                  "refId": "A",
                },
                "queryType": "alerting",
                "refId": "A",
                "relativeTimeRange": {
                  "from": 1000,
                  "to": 2000,
                },
              },
            ],
            "exec_err_state": "Error",
            "is_paused": false,
            "namespace_uid": "uuid020c61ef",
            "no_data_state": "NoData",
            "rule_group": "grafana-group-1",
            "title": "updated rule title",
            "uid": "4d7125fee983",
          },
          "labels": {
            "region": "nasa",
            "severity": "critical",
          },
        },
      ],
    },
    "headers": [
      [
        "content-type",
        "application/json",
      ],
      [
        "accept",
        "application/json, text/plain, */*",
      ],
    ],
    "method": "POST",
    "url": "http://localhost/api/ruler/grafana/api/v1/rules/uuid020c61ef?subtype=cortex",
  },
]
`;
